pandas/_libs/lib.pyx

```diff
@@ -1240,6 +1243,10 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     elif hasattr(value, 'dtype'):
         # this will handle ndarray-like
         # e.g. categoricals
+        # begin by checking if there is a tz attribute
```

Comment:
hmm, this is not the right place for this

Reply:
Ok so I thought about it again and prepared some tests.

<details>
 <summary>Tests</summary>

```python
import numpy as np
import pandas as pd
import pytz
import datetime
from pandas._libs.lib import infer_dtype

tests = [pd.Series([pd.Timestamp('2013-01-01', tz='UTC'), np.datetime64('NaT')]),
         pd.Series([pd.Timestamp('2013-01-01', tz='UTC')]),
         pd.Series([pd.Timestamp('2013-01-01', tz='America/Toronto'), pd.Timestamp('2013-01-01', tz='Europe/Paris')]),
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), np.datetime64('NaT')],
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), None],
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), datetime.datetime(2018, 1, 1, tzinfo=pytz.timezone('America/Toronto'))],
         np.array(datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc))]

for i, test in enumerate(tests):
    print(f'running test {i}')
    assert infer_dtype(test) == 'datetimetz'

```
</details>

And you are right I think the logical place is actually after the "except TypeError" for extension arrays and right before "_try_infer_map". By the way if you removed this modification of mine (where I check the "tz" attribute) then only the second test would actually fail.

Anyhow here is the code in the new place and all tests pass:

```python
        except TypeError:
            # This gets hit if we have an EA, since cython expects `values`
            #  to be an ndarray
            # before using try_infer_map check if there is a tz attribute
            # because try_infer_map cannot be used to differentiate dtypes
            # such as "datetime64[ns]" and "datetime64[ns, {timezone}]"
            if hasattr(value.dtype, 'tz'):
                return "datetimetz"
            value = _try_infer_map(value)
```
Btw if you are wondering why I have not modified "_try_infer_map" here is an explanation below.

<details>

<summary>why I have not changed _try_infer_map</summary>

So in the case of the second test the dtype is "datetime64[ns, UTC]" and the attribute "kind" of the dtype is "M" which is in the mapper "_TYPE_MAP"  so the value gets picked up by the function "_try_infer_map" returning "datetime64". 

Adding "datetime64[ns, UTC]"  to the mapper does not seem like a good solution to me since we would need to add every other timezones e.g. "datetime64[ns, Europe/Paris]" (I tried it and the same scenario happens).

Of course we could make some kind of rule in _try_infer_map I suppose (name of dtype starts with datetime64 etc.) but that does not seem very elegant to me either and does not seem to fit with the purpose of the function.

</details>

Reply:
you need to move this